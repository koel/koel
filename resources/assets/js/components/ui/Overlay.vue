<template>
  <dialog ref="el" :class="state.type" @cancel.prevent="onCancel" data-testid="overlay">
    <div class="wrapper">
      <SoundBars v-if="state.type === 'loading'"/>
      <icon v-if="state.type === 'error'" :icon="faCircleExclamation"/>
      <icon v-if="state.type === 'warning'" :icon="faWarning"/>
      <icon v-if="state.type === 'info'" :icon="faCircleInfo"/>
      <icon v-if="state.type === 'success'" :icon="faCircleCheck"/>

      <span class="message" v-html="state.message"/>
    </div>
  </dialog>
</template>

<script lang="ts" setup>
import { faCircleCheck, faCircleExclamation, faCircleInfo, faWarning } from '@fortawesome/free-solid-svg-icons'
import { defineAsyncComponent, reactive, ref } from 'vue'

const SoundBars = defineAsyncComponent(() => import('@/components/ui/SoundBars.vue'))

const el = ref<HTMLDialogElement>()

const state = reactive<OverlayState>({
  dismissible: false,
  type: 'loading',
  message: ''
})

const show = (options: Partial<OverlayState> = {}) => {
  Object.assign(state, options)
  el.value?.open || el.value?.showModal()
}

const hide = () => el.value?.close()
const onCancel = () => state.dismissible && hide()

defineExpose({ show, hide })
</script>

<style lang="scss" scoped>
dialog {
  border: 0;
  padding: 0;
  background: transparent;

  &::backdrop {
    background: rgba(0, 0, 0, 0.8);
  }

  .wrapper {
    display: flex;
    align-items: baseline;
    justify-content: center;
    gap: 6px;
  }

  &.error {
    color: var(--color-red);
  }

  &.success {
    color: var(--color-green);
  }

  &.info {
    color: var(--color-blue);
  }

  &.loading {
    color: var(--color-text-secondary);
  }

  &.warning {
    color: var(--color-orange);
  }
}
</style>
