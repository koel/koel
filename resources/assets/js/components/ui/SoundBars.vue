<template>
  <i class="relative flex gap-1 content-between w-[13px] aspect-square">
    <span v-for="i in 3" :key="i" />
  </i>
</template>

<style lang="postcss" scoped>
span {
  @apply w-[3px] h-full bg-current rounded-full origin-bottom;
  animation: bounce 2.2s ease-out infinite alternate;

  &:nth-of-type(2) {
    animation-delay: -2.2s;
  }

  &:nth-of-type(3) {
    animation-delay: -3.7s;
  }
}

@keyframes bounce {
  10% {
    transform: scaleY(0.2);
  }

  40% {
    transform: scaleY(0.7);
  }

  60% {
    transform: scaleY(0.5);
  }

  75% {
    transform: scaleY(1);
  }

  100% {
    transform: scaleY(0.4);
  }
}
</style>
