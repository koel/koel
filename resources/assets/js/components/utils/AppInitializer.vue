<template>
  <slot />
</template>

<script lang="ts" setup>
import { onMounted } from 'vue'
import { useErrorHandler, useOverlay } from '@/composables'
import { commonStore, preferenceStore as preferences } from '@/stores'
import { socketListener, socketService, uploadService } from '@/services'

const emits = defineEmits<{
  (e: 'success'): void
  (e: 'error', err: unknown): void
}>()

const { showOverlay, hideOverlay } = useOverlay()

/**
 * Request for notification permission if it's not provided and the user is OK with notifications.
 */
const requestNotificationPermission = async () => {
  if (preferences.show_now_playing_notification && window.Notification && window.Notification.permission !== 'granted') {
    preferences.show_now_playing_notification = await window.Notification.requestPermission() === 'denied'
  }
}

onMounted(async () => {
  showOverlay({ message: 'Just a little patienceâ€¦' })

  try {
    await commonStore.init()

    await requestNotificationPermission()

    window.addEventListener('beforeunload', (e: BeforeUnloadEvent) => {
      if (uploadService.shouldWarnUponWindowUnload() || preferences.confirm_before_closing) {
        e.preventDefault()
        e.returnValue = ''
      }
    })

    await socketService.init() && socketListener.listen()

    emits('success')
  } catch (error: unknown) {
    useErrorHandler().handleHttpError(error)
    emits('error', error)
  } finally {
    hideOverlay()
  }
})
</script>

<style lang="postcss" scoped>

</style>
