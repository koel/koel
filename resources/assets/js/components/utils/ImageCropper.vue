<template>
  <div class="w-full h-full fixed top-0 left-0 flex items-center justify-center z-[99] bg-black/70">
    <div class="relative max-w-full max-h-full rounded-md flex">
      <Cropper
        ref="cropper"
        :max-width="config.maxWidth"
        :min-width="config.minWidth"
        :src="source"
        :stencil-props="{ aspectRatio: 1 }"
      />
      <div class="fixed top-6 right-6 flex flex-1 gap-2">
        <Btn success @click.prevent="crop">Crop</Btn>
        <Btn transparent @click.prevent="emits('cancel')">Cancel</Btn>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, toRefs } from 'vue'
import { Cropper } from 'vue-advanced-cropper'
import 'vue-advanced-cropper/dist/style.css'

import Btn from '@/components/ui/form/Btn.vue'

const props = withDefaults(defineProps<{
  source?: string | null
  config?: {
    minWidth: number
    maxWidth?: number
  }
}>(), {
  source: null,
  config: () => ({
    minWidth: 192,
  }),
})

const emits = defineEmits<{
  (e: 'crop', result: string): void
  (e: 'cancel'): void
}>()
const { source, config } = toRefs(props)
const cropper = ref<typeof Cropper>()

const crop = () => emits('crop', cropper.value?.getResult().canvas.toDataURL())
</script>
