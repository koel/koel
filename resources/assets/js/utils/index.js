export * from "./alerts";
export * from "./filters";
export * from "./formatters";
export * from "./supports";
export * from "./common";
export * from "./$";
