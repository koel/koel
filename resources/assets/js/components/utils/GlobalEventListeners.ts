/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent, onMounted } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow } from '@/utils'
import { useDialogBox, useMessageToaster, useRouter } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    let toastSuccess: ReturnType<typeof useMessageToaster>['toastSuccess']
    let showConfirmDialog: ReturnType<typeof useDialogBox>['showConfirmDialog']
    let go: ReturnType<typeof useRouter>['go']

    onMounted(() => {
      toastSuccess = useMessageToaster().toastSuccess
      showConfirmDialog = useDialogBox().showConfirmDialog
      go = useRouter().go
    })

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
