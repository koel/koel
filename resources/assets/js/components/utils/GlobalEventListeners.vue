<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { DialogBoxKey, MessageToasterKey, RouterKey } from '@/symbols'

const toaster = requireInjection(MessageToasterKey)
const dialog = requireInjection(DialogBoxKey)
const router = requireInjection(RouterKey)

eventBus.on({
  PLAYLIST_DELETE: async (playlist: Playlist) => {
    if (await dialog.value.confirm(`Delete the playlist "${playlist.name}"?`)) {
      await playlistStore.delete(playlist)
      toaster.value.success(`Playlist "${playlist.name}" deleted.`)
      router.go('home')
    }
  },

  PLAYLIST_FOLDER_DELETE: async (folder: PlaylistFolder) => {
    if (await dialog.value.confirm(`Delete the playlist folder "${folder.name}"?`)) {
      await playlistFolderStore.delete(folder)
      toaster.value.success(`Playlist folder "${folder.name}" deleted.`)
      router.go('home')
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  LOG_OUT: async () => {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  }
})

router.onRouteChanged(() => {
  // Hide the extra panel away if a main view is triggered on mobile.
  if (isMobile.phone) {
    preferenceStore.showExtraPanel = false
  }
})
</script>
