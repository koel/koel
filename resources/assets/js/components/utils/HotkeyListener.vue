<template>
  <GlobalEvents
    @keydown.space="togglePlayback"
    @keydown.j="playNext"
    @keydown.k="playPrev"
    @keydown.f="search"
    @keydown.l="toggleLike"
  />
</template>

<script lang="ts" setup>
import { GlobalEvents } from 'vue-global-events'
import { eventBus } from '@/utils'
import { playbackService, socketService } from '@/services'
import { favoriteStore, queueStore } from '@/stores'

const togglePlayback = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, textarea, button, select') &&
    !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  e.preventDefault()
  playbackService.toggle()

  return false
}

/**
 * Play the previous song when user presses K.
 */
const playPrev = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playPrev()

  return false
}

/**
 * Play the next song when user presses J.
 */
const playNext = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playNext()

  return false
}

/**
 * Put focus into the search field when user presses F.
 */
const search = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, select, textarea') && !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  if (e.metaKey || e.ctrlKey) {
    return true
  }

  e.preventDefault()
  eventBus.emit('FOCUS_SEARCH_FIELD')

  return false
}

/**
 * Like/unlike the current song when use presses L.
 */
const toggleLike = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  if (!queueStore.current) {
    return false
  }

  favoriteStore.toggleOne(queueStore.current)
  socketService.broadcast('SOCKET_SONG', queueStore.current)

  return false
}
</script>
