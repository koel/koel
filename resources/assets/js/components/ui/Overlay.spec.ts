import UnitTestCase from '@/__tests__/UnitTestCase'
import { expect, it } from 'vitest'
import { eventBus } from '@/utils'
import SoundBars from '@/components/ui/SoundBars.vue'
import Overlay from './Overlay.vue'

new class extends UnitTestCase {
  private async renderComponent (type: OverlayState['type'] = 'loading') {
    const rendered = this.render(Overlay, {
      global: {
        stubs: {
          SoundBars
        }
      }
    })

    eventBus.emit('SHOW_OVERLAY', {
      type,
      message: 'Look at me now'
    })

    await this.tick()

    return rendered
  }

  protected test () {
    it.each<[OverlayState['type']]>([
      ['loading'],
      ['success'],
      ['info'],
      ['warning'],
      ['error']
    ])('renders %s type', async type => {
      const { getByTestId, html } = await this.renderComponent(type)

      expect(html()).toMatchSnapshot()
      expect((getByTestId('overlay') as HTMLDialogElement).open).toBe(true)
    })

    it('closes', async () => {
      const { getByTestId } = await this.renderComponent()

      eventBus.emit('HIDE_OVERLAY')
      expect((getByTestId('overlay') as HTMLDialogElement).open).toBe(false)
    })
  }
}
