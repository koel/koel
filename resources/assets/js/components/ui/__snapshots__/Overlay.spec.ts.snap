// Vitest Snapshot v1

exports[`renders error type 1`] = `
<div id="overlay" class="error overlay" data-testid="overlay">
  <div class="display">
    <!--v-if--><br data-testid="icon" icon="[object Object]">
    <!--v-if-->
    <!--v-if-->
    <!--v-if--><span class="message">Look at me now</span>
  </div>
  <!--v-if-->
</div>
`;

exports[`renders info type 1`] = `
<div id="overlay" class="info overlay" data-testid="overlay">
  <div class="display">
    <!--v-if-->
    <!--v-if-->
    <!--v-if--><br data-testid="icon" icon="[object Object]">
    <!--v-if--><span class="message">Look at me now</span>
  </div>
  <!--v-if-->
</div>
`;

exports[`renders loading type 1`] = `
<div id="overlay" class="loading overlay" data-testid="overlay">
  <div class="display"><i data-v-47e95701=""><span data-v-47e95701=""></span><span data-v-47e95701=""></span><span data-v-47e95701=""></span></i>
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if--><span class="message">Look at me now</span>
  </div>
  <!--v-if-->
</div>
`;

exports[`renders success type 1`] = `
<div id="overlay" class="success overlay" data-testid="overlay">
  <div class="display">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if--><br data-testid="icon" icon="[object Object]"><span class="message">Look at me now</span>
  </div>
  <!--v-if-->
</div>
`;

exports[`renders warning type 1`] = `
<div id="overlay" class="warning overlay" data-testid="overlay">
  <div class="display">
    <!--v-if-->
    <!--v-if--><br data-testid="icon" icon="[object Object]">
    <!--v-if-->
    <!--v-if--><span class="message">Look at me now</span>
  </div>
  <!--v-if-->
</div>
`;
